### PR TITLE
alloc: add some try_* methods Rust-for-Linux needs

### PR DESCRIPTION
**Pull request by Xuanwo**
This PR is a follow-up to the https://github.com/rust-lang/rust/pull/86938.

Part work of https://github.com/rust-lang/rust/issues/86942

---

> Based off of https://github.com/Rust-for-Linux/linux/commit/487d7578bd036ba57b0998d7af69e0bdcc595da8

Has been rebased onto rust-lang's master.

I will figure out all CI problems before request review.

### CODE REVIEW

**Comment by Ericson2314** [comment]
Thank you!!!

File: library/std/src/collections/hash/map.rs [hunk 1]
```diff
@@ -3011,10 +3011,10 @@ fn map_entry<'a, K: 'a, V: 'a>(raw: base::RustcEntry<'a, K, V>) -> Entry<'a, K,
 #[inline]
 pub(super) fn map_try_reserve_error(err: hashbrown::TryReserveError) -> TryReserveError {
     match err {
-        hashbrown::TryReserveError::CapacityOverflow => {
+        hashbrown::TryReserveErrorKind::CapacityOverflow => {
```
Comment (Xuanwo) [inline library/std/src/collections/hash/map.rs:3014]:
```suggestion
        hashbrown::TryReserveError::CapacityOverflow => {
```

File: library/std/src/collections/hash/map.rs [hunk 2]
```diff
@@ -3011,10 +3011,10 @@ fn map_entry<'a, K: 'a, V: 'a>(raw: base::RustcEntry<'a, K, V>) -> Entry<'a, K,
...
             TryReserveErrorKind::CapacityOverflow.into()
         }
-        hashbrown::TryReserveError::AllocError { layout } => {
+        hashbrown::TryReserveErrorKind::AllocError { layout } => {
```
Comment (Xuanwo) [inline library/std/src/collections/hash/map.rs:3017]:
```suggestion
        hashbrown::TryReserveError::AllocError { layout } => {
```

**Comment by Xuanwo** [comment]
ping @Ericson2314, this PR basically works now. Can you take a quick to confirm that I'm on the right track?

**Comment by Xuanwo** [comment]
I can't reproduce the failed at my side via: `./x.py test src/test/run-make-fulldeps --stage 2`.

**Comment by Ericson2314** [comment]
@Xuanwo do you think you could preserve the old history splitting into multiple commits, such that each work? Also remember you need to test with `--cfg no_global_oom_handling` too, that is where the failure is coming from.

**Comment by Xuanwo** [comment]
The history is so complex that I can't preserve them simply (the base is so different).  My plan is to squash  all old commits into one which authored with your name and keep my changes as you suggested. What do you think?

**Comment by Ericson2314** [comment]
@Xuanwo the reason I did the split was because some of the later commits might introduce performance regressions. It was something that came up in the meetings.

**Comment by Ericson2314** [comment]
I will take a brief look at the rebase.

**Comment by Xuanwo** [comment]
@Ericson2314 OK, let me try it again. 

Is there any convenient way to switch the base from Rust-for-Linux to rust-lang? I spent a lot time to fix the conflict :(

**Comment by Ericson2314** [comment]
@Xuanwo give me like 10 minutes to see if I can do it on your behalf. I have done a lot of rebase I think I can port over your conflict resolutions OK.

(I also feel it's a bit on me because I didn't mention that preserving history is good from the get-go.)

**Comment by Ericson2314** [comment]
Unfortunately GitHub doesn't let one change the base branch of PRs.

**Comment by Xuanwo** [comment]
> Unfortunately GitHub doesn't let one change the base branch of PRs.

Yes, we have to do it via git.

**Comment by Ericson2314** [comment]
@Xuanwo OK I did the rebase and forced pushed. It can now be compared to your branch, since they share your base branch.

What I recommend doing is an interactive on that same base commit (d594910a2da12f158477b4c7281716f535cfa3de), but where you edit each commit of mine, and pull over the relevant fixes from yours for that commit.

Now that can glance the diff showing your changes, they look good :). Just make sure `--cfg no_global_oom_handling` works as you pull over those changes, and then I think we're good!

**Comment by Xuanwo** [comment]
I can't continue this work due to lack of time recently, let's close this PR for now, thanks for the help!

**Pull request closed without merging**